lib/services/authentication.go

```diff
@@ -19,13 +19,401 @@ limitations under the License.
 package services
 
 import (
+	"bytes"
+	"crypto/ecdsa"
+	"crypto/x509"
 	"encoding/json"
 	"fmt"
+	"time"
 
+	"github.com/gravitational/teleport/api/constants"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/defaults"
 	"github.com/gravitational/teleport/lib/utils"
+
 	"github.com/gravitational/trace"
+	"github.com/pquerna/otp/totp"
+	"github.com/tstranex/u2f"
+	"golang.org/x/crypto/bcrypt"
 )
 
+// AuthPreference defines the authentication preferences for a specific
+// cluster. It defines the type (local, oidc) and second factor (off, otp, oidc).
+// AuthPreference is a configuration resource, never create more than one instance
+// of it.
+type AuthPreference interface {
+	// Resource provides common resource properties.
+	Resource
+
+	// GetType gets the type of authentication: local, saml, or oidc.
+	GetType() string
+	// SetType sets the type of authentication: local, saml, or oidc.
+	SetType(string)
+
+	// GetSecondFactor gets the type of second factor: off, otp or u2f.
+	GetSecondFactor() string
+	// SetSecondFactor sets the type of second factor: off, otp, or u2f.
+	SetSecondFactor(string)
+
+	// GetConnectorName gets the name of the OIDC or SAML connector to use. If
+	// this value is empty, we fall back to the first connector in the backend.
+	GetConnectorName() string
+	// SetConnectorName sets the name of the OIDC or SAML connector to use. If
+	// this value is empty, we fall back to the first connector in the backend.
+	SetConnectorName(string)
+
+	// GetU2F gets the U2F configuration settings.
+	GetU2F() (*U2F, error)
+	// SetU2F sets the U2F configuration settings.
+	SetU2F(*U2F)
+
+	// CheckAndSetDefaults sets and default values and then
+	// verifies the constraints for AuthPreference.
+	CheckAndSetDefaults() error
+
+	// String represents a human readable version of authentication settings.
+	String() string
+}
+
+// NewAuthPreference is a convenience method to to create AuthPreferenceV2.
+func NewAuthPreference(spec AuthPreferenceSpecV2) (AuthPreference, error) {
+	pref := AuthPreferenceV2{
+		Kind:    KindClusterAuthPreference,
+		Version: V2,
+		Metadata: Metadata{
+			Name:      MetaNameClusterAuthPreference,
+			Namespace: defaults.Namespace,
+		},
+		Spec: spec,
+	}
+
+	if err := pref.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &pref, nil
+}
+
+// DefaultAuthPreference returns the default authentication preferences.
+func DefaultAuthPreference() AuthPreference {
+	return &AuthPreferenceV2{
+		Kind:    KindClusterAuthPreference,
+		Version: V2,
+		Metadata: Metadata{
+			Name:      MetaNameClusterAuthPreference,
+			Namespace: defaults.Namespace,
+		},
+		Spec: AuthPreferenceSpecV2{
+			Type:         constants.Local,
+			SecondFactor: constants.OTP,
+		},
+	}
+}
+
+// AuthPreferenceV2 implements AuthPreference.
+type AuthPreferenceV2 struct {
```

Comment:
Why is this moving back?

Reply:
I moved the entire file back to `lib/services` because it was no longer needed by `api/client`. It may need to be moved back once we convert the http endpoint to gRPC. My thought process during the dependency refactors has been to move everything that is not needed by the api back to services, now that I have more context and understanding of the code. I think I moved a lot of things that should have stayed in services.